File: packages/node/integration.mjs
import { randomUUID } from 'node:crypto';
import { LLMtrack } from './dist/index.js';

const apiKey = process.env.LLMTRACK_API_KEY;
if (!apiKey) {
  console.error('FAIL setup: LLMTRACK_API_KEY is required');
  process.exit(1);
}

let failures = 0;
async function check(name, operation) {
  try { await operation(); console.log(`PASS ${name}`); }
  catch (error) { failures++; console.log(`FAIL ${name}: ${error?.message ?? error}`); }
}
const event = {provider:'openai', model:'gpt-4o-mini', promptTokens:2, completionTokens:1, reasoningTokens:1, feature:'sdk-integration'};
const client = new LLMtrack({apiKey});

await check('happy path with reasoning tokens', async()=>{ const result=await client.trackSync(event); if(!result) throw Error('empty response'); });
await check('invalid key is reported and track never throws', async()=>{
  const error = new Promise((resolve,reject)=>{ const timer=setTimeout(()=>reject(Error('onError timeout')),7000); new LLMtrack({apiKey:'invalid-integration-key',maxRetries:1,onError:e=>{clearTimeout(timer);resolve(e);}}).track(event); });
  const result=await error; if(result.code!=='INVALID_API_KEY') throw Error(`unexpected ${result.code}`);
});
await check('free-plan visibility warning', async()=>{
  const warnings=[]; const c=new LLMtrack({apiKey,environment:`sdk-mismatch-${randomUUID()}`,onWarning:w=>warnings.push(w)});
  await c.trackSync(event); if(!warnings.some(w=>w.code==='NOT_DASHBOARD_VISIBLE')) throw Error('API did not return dashboard_visible=false; use a free-plan test key bound to another environment');
});
await check('unknown-model pricing warning', async()=>{
  const warnings=[]; const c=new LLMtrack({apiKey,onWarning:w=>warnings.push(w)});
  await c.trackSync({...event,model:`unknown-integration-${randomUUID()}`}); if(!warnings.some(w=>w.code==='UNKNOWN_MODEL')) throw Error('missing UNKNOWN_MODEL warning');
});
await check('oversized metadata is rejected client-side', async()=>{ try{await client.trackSync({...event,metadata:{value:'x'.repeat(8193)}});}catch(e){if(e.code==='INVALID_PAYLOAD')return;}throw Error('expected INVALID_PAYLOAD'); });
await check('negative tokens are rejected client-side', async()=>{ try{await client.trackSync({...event,promptTokens:-1});}catch(e){if(e.code==='INVALID_PAYLOAD')return;}throw Error('expected INVALID_PAYLOAD'); });
await check('same idempotency key is a successful duplicate', async()=>{const key=randomUUID();await client.trackSync({...event,idempotencyKey:key});const second=await client.trackSync({...event,idempotencyKey:key});if(second.duplicate!==true)throw Error('second response was not duplicate');});
await check('unreachable network failure is reported without throwing', async()=>{
  const error=new Promise((resolve,reject)=>{const timer=setTimeout(()=>reject(Error('onError timeout')),3000);new LLMtrack({apiKey,baseUrl:'http://127.0.0.1:1',timeoutMs:250,maxRetries:1,onError:e=>{clearTimeout(timer);resolve(e);}}).track(event);});
  const result=await error;if(result.code!=='NETWORK_ERROR')throw Error(`unexpected ${result.code}`);
});
process.exitCode = failures ? 1 : 0;

File: packages/python/integration.py
"""Opt-in live checks. Requires a dedicated LLMTRACK_API_KEY."""
import asyncio
import os
import uuid

from llmtrack_sdk import LLMtrack, LLMtrackError

API_KEY = os.environ.get("LLMTRACK_API_KEY")
if not API_KEY:
    raise SystemExit("FAIL setup: LLMTRACK_API_KEY is required")
EVENT = dict(provider="openai", model="gpt-4o-mini", prompt_tokens=2, completion_tokens=1,
             reasoning_tokens=1, feature="sdk-integration")
failures = 0

async def check(name, operation):
    global failures
    try:
        await operation()
        print(f"PASS {name}")
    except Exception as exc:
        failures += 1
        print(f"FAIL {name}: {exc}")

async def callback_error(client):
    loop = asyncio.get_running_loop()
    done = loop.create_future()
    def receive(error):
        loop.call_soon_threadsafe(lambda: None if done.done() else done.set_result(error))
    client.on_error = receive
    client.track(**EVENT)
    return await asyncio.wait_for(done, 7)

async def main():
    client = LLMtrack(api_key=API_KEY)
    await check("happy path with reasoning tokens", lambda: client.track_sync(**EVENT))
    async def invalid():
        error = await callback_error(LLMtrack(api_key="invalid-integration-key", max_retries=1))
        assert error.code == "INVALID_API_KEY", error.code
    await check("invalid key is reported and track never raises", invalid)
    async def visibility():
        warnings=[]; c=LLMtrack(api_key=API_KEY, environment=f"sdk-mismatch-{uuid.uuid4()}", on_warning=warnings.append)
        await c.track_sync(**EVENT)
        assert any(w.code == "NOT_DASHBOARD_VISIBLE" for w in warnings), "use a free-plan test key bound to another environment"
    await check("free-plan visibility warning", visibility)
    async def pricing():
        warnings=[]; c=LLMtrack(api_key=API_KEY, on_warning=warnings.append)
        await c.track_sync(**{**EVENT,"model":f"unknown-integration-{uuid.uuid4()}"})
        assert any(w.code == "UNKNOWN_MODEL" for w in warnings), "missing UNKNOWN_MODEL warning"
    await check("unknown-model pricing warning", pricing)
    async def invalid_payload(**changes):
        try: await client.track_sync(**{**EVENT,**changes})
        except LLMtrackError as exc:
            assert exc.code == "INVALID_PAYLOAD"; return
        raise AssertionError("expected INVALID_PAYLOAD")
    await check("oversized metadata is rejected client-side", lambda: invalid_payload(metadata={"value":"x"*8193}))
    await check("negative tokens are rejected client-side", lambda: invalid_payload(prompt_tokens=-1))
    async def duplicate():
        key=str(uuid.uuid4()); await client.track_sync(**EVENT,idempotency_key=key)
        second=await client.track_sync(**EVENT,idempotency_key=key); assert second["duplicate"] is True
    await check("same idempotency key is a successful duplicate", duplicate)
    async def network():
        error=await callback_error(LLMtrack(api_key=API_KEY,base_url="http://127.0.0.1:1",timeout_ms=250,max_retries=1))
        assert error.code == "NETWORK_ERROR", error.code
    await check("unreachable network failure is reported without raising", network)

asyncio.run(main())
raise SystemExit(1 if failures else 0)
